app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/ResolveCategoryAggregation.php

```diff
@@ -0,0 +1,167 @@
+
+/**
+ * Resolve category aggregation.
+ */
+class ResolveCategoryAggregation
+{
+    /**
+     * @var string
+     */
+    public const CATEGORY_BUCKET = 'category_bucket';
+
+    /**
+     * @var string
+     */
+    private const BUCKETS = 'buckets';
+
+    /**
+     * @var AggregationFactory
+     */
+    private $aggregationFactory;
+
+    /**
+     * @var BucketFactory
+     */
+    private $bucketFactory;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @var CategoryRepository
+     */
+    private $categoryRepository;
+
+    /**
+     * @var array
+     */
+    private $resolvedChildrenIds = [];
+
+    /**
+     * @param AggregationFactory $aggregationFactory
+     * @param BucketFactory $bucketFactory
+     * @param StoreManagerInterface $storeManager
+     * @param CategoryRepository $categoryRepository
+     */
+    public function __construct(
+        AggregationFactory $aggregationFactory,
+        BucketFactory $bucketFactory,
+        StoreManagerInterface $storeManager,
+        CategoryRepository $categoryRepository
+    ) {
+        $this->aggregationFactory = $aggregationFactory;
+        $this->bucketFactory = $bucketFactory;
+        $this->storeManager = $storeManager;
+        $this->categoryRepository = $categoryRepository;
+    }
+
+    /**
+     * Get resolved category aggregation.
+     *
+     * @param array $categoryFilter
+     * @param Bucket[] $bucketList
+     *
+     * @return Aggregation
+     */
+    public function getResolvedCategoryAggregation(array $categoryFilter, array $bucketList): Aggregation
+    {
+        $categoryBucket = $bucketList[self::CATEGORY_BUCKET] ?? [];
+        $values = $categoryBucket->getValues();
+        $condition = array_key_first($categoryFilter);
+        $searchableCategoryValue = $categoryFilter[$condition] ?? 0;
+
+        if (!$searchableCategoryValue || empty($values)) {
+            return $this->aggregationFactory->create([self::BUCKETS => $bucketList]);
+        }
+
+        $resolvedBucketList = $bucketList;
+
+        try {
+            if (!is_array($searchableCategoryValue)) {
+                $resolvedValues = $this->getValidCategories((int) $searchableCategoryValue, $values);
+            } else {
+                $categoryResolvedValues = [];
+
+                foreach ($searchableCategoryValue as $categoryId) {
+                    $categoryResolvedValues[] = $this->getValidCategories((int) $categoryId, $values);
+                }
+
+                $resolvedValues = call_user_func_array('array_merge', $categoryResolvedValues);;
+            }
+        } catch (NoSuchEntityException $e) {
+            return $this->aggregationFactory->create([self::BUCKETS => $bucketList]);
+        }
+
+        $resolvedCategoryBucket = $this->bucketFactory->create(
+            [
+                'name' => self::CATEGORY_BUCKET,
+                'values' => $resolvedValues
+            ]
+        );
+
```

Comment:
Please remove redundant lines

```suggestion
        $resolvedCategoryBucket = $this->bucketFactory->create(
            [
                'name' => self::CATEGORY_BUCKET,
                'values' => $resolvedValues
            ]
        );
```